Feature: Chat Filter Logs

Road-block opened the pull request:
- Adds a tab to show the last 50 blocked messages for each active filter (selectable from a dropdown at the top)
- Moved [Options] tab to tab4 so it is again at the end.
- AddToLog function stores the last 50 messages for each active filter, pruning older records so the storage remains manageable.
  - pruning old records only happens out of combat to avoid re-indexing the store under combat
  - data driven gui updates also only happen if the gui is visible on the correct tab + filter selected
- M.ShowFilterLog (filter_id) API to open directly to a filter log (from other widgets or cmdline)
  - Add a Logs button to filter edit frame that opens directly to that filter's log
  - Make the infotext clickable as well with the same function.

Road-block committed:
store the last 50 chat filter hits to a per filter log
gui: add a filter logs tab that displays the last 50 blocked messages for each active filter
gui: move options to tab 4

Road-block commented:
Looks like this 
<img width="930" height="539" alt="chatfilters_gui" src="https://github.com/user-attachments/assets/29ff1554-2299-411d-91ed-79e31e9b8c00" />

Test package available [here](https://github.com/Road-block/GlobalIgnoreList/releases/latest)

Closes https://github.com/IceyPop/GlobalIgnoreList/issues/2 if accepted.

Road-block committed:
Merge branch 'main' into chat_filter_logs

Road-block committed:
fix /gi history cmd
add mists interface number to toc

Road-block committed:
add a Logs button to the filter edit page to open directly to the filter logs
make the filter info text clickable to also open directly to that filter's logs
add an M.ShowFilterLog (filterid) API to be used for opening gui directly to that filter log

Road-block commented:
<img width="933" height="550" alt="image" src="https://github.com/user-attachments/assets/5569e7d2-4974-44cf-90e8-cb88bd094990" />

IceyPop commented:
Hey there I know we kind of got out of sync in terms of our development.  I had that history stuff committed as an alpha as mentioned, but I also had a half done UI too.  So I spent a little time today cleaning up what I had done and committed to the alpha chain because it seems like we're both working on basically the same things independently lol!

Normally I wouldn't intend to commit anything until I think its mostly complete but I also want you to have a look at what I am doing too so we can coordinate better.  It sounds like we're pretty close to doing the same things I just didn't make mine on a new tab

Road-block commented:
Unless you have a more recent alpha push than the one from 5-ish days ago I saw that and I have resolved all conflicts between the two branches.
This should be apparent by a [no conflicts with base branch] status on this PR.
This means I incorporated your changes and merged them together with mine so that both "feature sets" are present.
For example your `/gui history` command line handler and other fixes you had here and there, as well the gui for it .

The only difference from your code is with naming (yours filterHistory, mine filterLogs) but that's a trivial change which I can swap over to your choice if that helps.

Other than that, actual changes are filter records holding more information than *just* the message and a fix for your command handler.

On the naming convention, I named the filter history "Logs" because unlike the command handler which would only show a snapshot at the time it's called the tab shows a real-time view of filter hits as they're coming in, so I figured Log is more appropriate.

_The command handler would also require some way for the player to know which filter ids are active and have records, so I figured it was more of a debug tool for a developer not really a UX solution for players. To make it that it would at the very least need_ `/gui history` _with no filter id parameter to return available ids._

I didn't repost that message but the test package (for in-game testing) has been kept updated at the same place [here](https://github.com/Road-block/GlobalIgnoreList/releases/latest)

Road-block commented:
Oh I see a new alpha, I will check it out :)

Road-block commented:
I like this implementation too.
Since there is not much point in merging between new tab and new tab2 frame my gui implementation is probably obsolete.

Small changes I would make to your current alpha implementation.
1. Cosmetic: "Your chat filters have blocked a total of N ~~spammers~~ messages". Same for the other instances of this info text on filters; "This filter has blocked N ~~spammers~~ messages". *I don't think either of them counts unique senders.*
2. I could bring over the "click info text" and/or extra "history" button next to Test as alternative options for navigating to the block frame from the edit frame.
3. On the block frame itself, I would still like to have the channel and timestamp information available somewhere so maybe bring over the mouseover tooltip for those? (I also had only 2 fields visible and the other 2 infos on a tooltip, just different choice of fields)
4. I think 20 records is a bit low for really active filters to catch false positives (my LFG spam filter blocks 100 messages/minute approx) and conversely no timestamp is felt more on inactive filters where you have no way to know what version of the filter blocked the message (if you have edited it between hits). So I still think that storing an array with data is better than storing a concatenated string as it's more future proof and lets you refine presentation easier.
5. There is an unintended consequence (I think) to the "Automatically update default Chat Filters". Since players are not prevented from editing those themselves, their updates are lost by addon updates currently with that option enabled. It should either prevent editing of built-in filters or some way to detect if it has been user edited?

Let me know what you think.

Road-block committed:
Merge branch 'alpha' of https://github.com/IceyPop/GlobalIgnoreList into chat_filter_logs

IceyPop commented:
Great feedback!

I have been going back and forth on how to handle the history collection.  For example, the time when someone will be in combat while having the most blocked text will be when in a town beating on a target dummy.  When outside of town fighting, the text volume will be much much lower.  So I went to the main city on retail and watched someone hit on a dummy to see how they behave.  This person hit the thing for maybe 5-10 minutes straight!  If they had filters turned on for the usual suspects, they could end up collecting a crazy amount of data if its only pruned out of combat.  I am torn on if its better to do it outside or keep the number lower and do it always, just in the most efficient way possible.

So that led to a table data test which is how I ended up with the string concat version.  I do think the way you're doing it is way better and much more elegant...  But the reason I didn't change to that (yet) was because my tests showed that string concat was 15% faster.  And given I feel this is a time critical place in the code, I sided with the 15% faster method.  But maybe I am overthinking it?  I can try to share the code I tested using a lua interpreter on Windows 11 (not WoW):

[test.txt](https://github.com/user-attachments/files/22873707/test.txt)

I'll add those other details as you have recommended to the history UI after I try to convert the UI to use buttons (I had some notes in the code I sort of coded myself into a wall by using only fontstrings on a scrolling frame without buttons).

The history command line was just meant to be a debug tool, I was planning to remove it before release.  But if you like it we could the filter # somewhere in the UI.  I do think it will get even more confusing though if we do, because eventually I'll probably add sortable columns to the chat filter list like I have in the ignore list.  So for now my plan is to remove it before next release.

Road-block commented:
There is a way to implement the cache that does not cause re-indexing at all.
The side-effect of it is that the order of the messages is not uniform. 
I thought about doing it that other way which sidesteps the whole pruning logic entirely but then you can't really watch a live log with consistent behavior (new messages coming in at the top or at the bottom)

```lua
local cache = {}
local idx, maxid = 1, 50
function addToCache(message)
  if idx == 50 then idx = 1 end
  cache[idx] = message
  idx = idx + 1
end
```
As a trivial example. This keeps the cache at 50 records and never re-indexes as when it gets to the max id it simply starts overwriting records from 1 again.
But obviously as said earlier it means that there are "breaks" in the order whenever it starts overwriting.

Road-block commented:
Just to say though that your test is not entirely _fair_ 😆 
Strings are interned in Lua so the concat operation only allocates memory once since none of the constituent parts change.

When storing the array you replace a bunch of string operations with simple lookups in UpdateBlockButtons as the relevant bits are as simple as
```lua
  local text = GlobalIgnoreDB.filterHistory[Tab2BlockFilter][count].m
  local player = GlobalIgnoreDB.filterHistory[Tab2BlockFilter][count].s
```
without needing to reverse engineer the string stored in the other place.

You make a good point about minutes long combat though, I'll think about it some more.

Road-block commented:
> convert the UI to use buttons (I had some notes in the code I sort of coded myself into a wall by using only fontstrings on a scrolling frame without buttons).

I saw the note, I don't think you need to convert them to buttons if all you want is to show a menu on the player field.
```lua
BlockScrollPlayer[idx]:SetScript("OnMouseUp", function(self,button)
	if button == "RightButton" then
		print("r-clicked "..count)
	end
end)
```
will do the trick.

Road-block committed:
minor localization cleanup
remove an unintended global
AddToBlockHistory API (uses an array)
refactor UpdateBlockButtons to use the new records
add tooltips to block message field for time and channel
add a click handler stub to player block field
add a button to navigate to block history from edit filter
clicking the infotext will also swap to the block history
fix the /gi history cmd

Road-block commented:
I re-targeted the PR towards your alpha branch instead of main to make comparisons easier.
Considering how much data a typical damage meter retains in that same 10minute period I really don't think storing the block history till next OOC is going to be a problem, let alone really popular chat addons like Chattynator.

I implemented most of the points from https://github.com/IceyPop/GlobalIgnoreList/pull/3#issuecomment-3394470215 (except 5)
<img width="928" height="547" alt="image" src="https://github.com/user-attachments/assets/7910c8f1-7414-4c4a-bfb9-62de55e950d2" />

I'm going to leave the PR alone 😄 

Take whatever you find useful, discard whatever doesn't fit your preference.

I'm happy with how the addon is shaping up, whatever implementation is chosen it'll be feature complete for my needs.

Road-block committed:
fix an errant localization entry

Road-block committed:
cleanup an unintended global and a nil reference in the chatfilter

Road-block committed:
fix some block history refresh issues
expand the block history player field context menu stub with actual functionality
restore the todo comment

Road-block commented:
Implemented an example menu for the player field of the block history instead of a stub.
<img width="927" height="551" alt="image" src="https://github.com/user-attachments/assets/a24e88ee-78f2-42ae-8b3d-7142b07cb43b" />
`Add to Ignore` menu item is context sensitive, it will be `Remove from Ignore` if the player is already ignored.

IceyPop commented:
Thank you for the submission but just noting I have already completed all the of the UI and right click stuff a couple of days ago when I last worked on things, so we're probably getting apart from each other again in terms of code.

Road-block commented:
> Thank you for the submission but just noting I have already completed all the of the UI and right click stuff a couple of days ago when I last worked on things, so we're probably getting apart from each other again in terms of code.

Hah, that's fine, just use your stuff 👍🏻

Road-block commented:
Since this has gone a bit stale I'll try solve that cache conundrum in a way that makes combat check unnecessary.

IceyPop commented:
Hey so sorry I have a big thing at work so I haven't had time in the past week.  I'll be back to normal in a couple of days.

If you want I can submit where I left off a week or so ago to the alpha branch when I get home from work (in about 2 hours).  But I don't want to add any confusion to what you're working on by doing that.

Either way I'll make sure I get your merges and a new release out sometime this weekend!

Road-block commented:
> If you want I can submit where I left off a week or so ago to the alpha branch when I get home from work (in about 2 hours). But I don't want to add any confusion to what you're working on by doing that.

If you want to push your stuff to alpha, I will respect the implementation, maybe add small fixes here and there that I had (in case they weren't picked up) talking about unintended globals, a nil reference etc, small cleanup stuff.

Only thing I want to check is if I can change the cache implementation to where it doesn't re-index and thus doesn't need a combat check at all.

IceyPop commented:
Its okay if you don't want to use/honor it, I think what we're doing is so close to each other!  I haven't touched it in about 9 or 10 days so I forget exactly where I left off, but I added the date to the history and the right click menu.  I stopped on a UI bug where the right click of the history text was still active below the UI window.  I didn't get a chance to convert it over to a table insert (still using a string) nor did I add the mouse over with the channel information.  So I am not sure if its worth just merging what you have or using this, but I am open to whatever! :)  I'll submit what I had up to about 10 days ago now.

Road-block commented:
I will resolve conflicts again but with a slight bias towards my stuff this time.
Main reason is a few artifacts have accumulated in the alpha and if I went the other way I'd have to refactor on top of the merge to fix them.
For example the current alpha code creates a new closure for the right-click menu every time the UpdateBlockButtons method is called.

Main UI difference from your latest alpha is going to be that I'm keeping the 2 column format (player | text) with the timestamp on the mouseover tooltip and I'm keeping the extra entries for the right-click menu according to your earlier notes in comments.

The scrollframe overflow will be fixed and the cache implementation has changed so that it never does a full re-index or grow larger than the set limit, so the combat check has been removed.

The slight duplication of code in remove single entry and remove all entries by player is intentional (I could have made the remove entries by player call the remove entry) but I kept them separate to limit the calls to UpdateBlockButtons.

Additionally with the new cache implementation, the history is iterated in 2 partial loops so the latest blocked messages are always coming in on top (for the gui) and on bottom (for the /gi history cmd)

Road-block committed:
refactor block history cache implementation
- entries are re-used when the entry limit is reached to avoid overflow / re-indexing
- track last entry id so gui or chat display of history order stays by entry age
right-click menu added to message field in addition to player, parse out menu entries to methods
deduplicate code in UpdateBlockButtons to better handle the new cache implementation gui refreshes
fix scrollframe overflow for block history
simplify all array bound checks to avoid repeated calls to the table length operator, new code will also gracefully handle arrays with 'holes'

Road-block committed:
Merge branch 'alpha' of https://github.com/IceyPop/GlobalIgnoreList into chat_filter_logs

Road-block commented:
Conflicts have been resolved and this can be cleanly merged into alpha branch (as it is right now)
Apart from the notes in the previous comment for additions/refactors the one thing that was "lost" in the merge is BlockScrollDate

If it's preferred for it to be a visible column rather than a tooltip entry it should be fairly simple to re-add after the merge at [UpdateBlockButton](https://github.com/Road-block/GlobalIgnoreList/blob/2777e024c29b9242859f4605d59cbd6c2126d1be/GIL_UI.lua#L420)
[M.UpdateBlockButtons](https://github.com/Road-block/GlobalIgnoreList/blob/2777e024c29b9242859f4605d59cbd6c2126d1be/GIL_UI.lua#L478) would need a similar minor edit adding the field resets and finally the tooltip entry and field/column widths adjusted.

<img width="927" height="550" alt="image" src="https://github.com/user-attachments/assets/ec01f743-28bb-4123-bd82-5358a5519b7b" />

Test package for the resulting addon if PR was merged [here](https://github.com/Road-block/GlobalIgnoreList/releases/latest)

IceyPop commented:
Hey sorry for slacking on this, I had to step away from WoW for a few weeks.  I completely butchered trying to merge since I am new to GitHub, but your idea for pruning and the table storage/mouseover in the block history is all in the main now.

IceyPop commented:
Closing (accepted but I accidentally committed from my local disk instead of merging your pull request sorry!)
